### ISSUE TITLE
Internal turbodbc exception after many executemany statements (Azure SQL Database)

### ISSUE THREAD

**Issue by sevetseh28** [blue-yonder/turbodbc#141]
- Python 3.6.4.
- Turbodbc 2.4.1
- Ubuntu 16.04
- Target database: Azure SQL Database
- Driver: Official Microsoft Driver 13.1

After performing many executemany statements with batches of 5000 rows I randomly get this exception:

```
2018-01-09 17:09:09,641 - dal.dal - ERROR - Traceback (most recent call last):
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/exceptions.py", line 50, in wrapper
    return f(*args, **kwds)
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/cursor.py", line 165, in executemany
    buffer.flush()
turbodbc_intern.Error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/c/IDATHA/Workspace/idatha_sbi/dal/dal.py", line 292, in load_wordcloud
    wordcloud_tuples_batch)
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/exceptions.py", line 52, in wrapper
    raise DatabaseError(str(e))
turbodbc.exceptions.DatabaseError
```

I already tried setting parameter_sets_to_buffer to >5000. I am pretty sure this is a turboODBC bug because wasn't happening before with another library like ceODBC.

Thank you!

**Comment by MathMagique** [comment]
That's unfortunate. Could you provide some Python code to reproduce the issue? Can you check how often `executemany()` works before it fails, assuming you use the same cursor? Depending on your code, just using one `executemany()` statement or closing the cursor from time to time might be a workaround.

Anyway, the Python code to reproduce this issue would be really useful.

**Comment by sevetseh28** [comment]
Hi, @MathMagique . I was using one cursor for all statements but I already tried one cursor per executemany and the same exception occurs.  Using only one executemany statement would be out of the question since I need to insert about 4 million tuples.

I think the exception is thrown randomly after some time, e.g. if I change the size of the batch to 10.000 the exception occurs at the 10th batch (100.000 rows inserted) but now it happened at 300.000 rows with a batch size of 6000.
Even if you look at my code, I close the connection and reopen it to retry insertion but it keeps throwing the exception even with a new connection. However, If I rerun the script it starts inserting with no problems until it throws it again at some point.

Here's a snippet of my loop:

     ```
   for wordcloud_tuples_batch in batch(wordcloud_tuples, **BATCH_SIZE**):
            cursor_azure.close()
            cursor_azure = self.cnxn_azure.cursor()
            commited = False
            while True:
                try:
                    start_time = time.time()
                    if source == TWITTER:
                        cursor_azure.executemany(
                            'INSERT INTO [staging].[tweets_wordcloud] (tweet_id, word, word_count) VALUES (?,?,?)',
                            wordcloud_tuples_batch)
                    elif source == FACEBOOK:
                        cursor_azure.executemany(
                            'INSERT INTO [staging].[facebook_objects_wordcloud] (object_id, word, word_count) VALUES (?,?,?)',
                            wordcloud_tuples_batch)
                    elif source == NEWS:
                        cursor_azure.executemany(
                            'INSERT INTO [staging].[news_wordcloud] (news_id, word, word_count) VALUES (?,?,?)',
                            wordcloud_tuples_batch)
                    total_inserted += len(wordcloud_tuples_batch)
                    self.logger.info(
                        '{} Wordcloud - {}/{} words inserted.  {}  secs inserting batch of {}'.format(source,
                                                                                                      total_inserted,
                                                                                                      len(
                                                                                                          wordcloud_tuples),
                                                                                                      time.time() - start_time,
                                                                                                      len(
                                                                                                          wordcloud_tuples_batch)))
                    self.cnxn_azure.commit()
                    self.logger.debug('Words Commited!')
                    commited = True
                except turbodbc.exceptions.DatabaseError:
                    self.logger.error(traceback.format_exc())
                    self.cnxn_azure.close()
                    self.cnxn_azure = self.get_azure_conn()
                    cursor_azure = self.cnxn_azure.cursor()
                    if commited:
                        self.logger.debug('Going to next batch')
                        break
                    else:
                        self.logger.debug('Retrying batch')
                        continue

                except Exception as e:
                    self.logger.error(traceback.format_exc())
                    cursor_azure.close()
                    exit(1)
                break
```

**Comment by MathMagique** [comment]
That's odd and will require a closer look. Thanks for the code!

Have you tried closing the connection in between? Also, the one `executemany()` approach could work if you pass in a generator instead of a list of tuples.

**Comment by sevetseh28** [comment]
I will try to reopen a connection between each batch. 
How would that generator thing be? I already have the 4 million tuples list in memory (wordcloud_tuples variable). What I do is retrieve batches from it (via a generator as you can see). Do you mean that I should pass the batch generator to the executemany statement?

Thanks for the help!

**Comment by MathMagique** [comment]
Yeah, now I realize that you already have everything in memory. Obviously, it was too late yesterday evening ;-).

In this case, I don't think you need a batch generator at all. Just try passing the entire `wordcloud_tuples` to `executemany()`. Turbodbc will automatically transfer data in batches. The batch size can be controlled via an option (see the [documentation](http://turbodbc.readthedocs.io/en/latest/pages/advanced_usage.html#buffered-parameter-sets)).

**Comment by sevetseh28** [comment]
Trying to executemany with the 4 million list throws:
```
2018-01-09 19:29:58,340 - dal.dal - ERROR - Traceback (most recent call last):
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/exceptions.py", line 50, in wrapper
    return f(*args, **kwds)
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/cursor.py", line 164, in executemany
    buffer.add_set(parameter_set)
turbodbc_intern.Error
```

Creating a new connection between each batch doesnt work either, I get a Login timeout eventually. :(

**Comment by sevetseh28** [comment]
It is now running with auto_commit set to True with the 4 million list passed to executemany and a parameter_sets_to_buffer set to 500. It has been running for about 45 minutes and no exception has been thrown. However it is extremely slow, only 350.000 rows have been inserted on that time.  So even if that works it is not a solution since I'm looking for performance as well and that is what turbodbc is good at right?
I'll try to spawn multiple processes and increase parameter_sets_to_buffer to 5000. I'll keep you posted.

PD: Doing multiples SELECT COUNT(*) in a short time I realized auto_commit=True actually performs a commit with a batch size lower than parameter_sets_to_buffer value so maybe that is why it is slowing down performance.

**Comment by MathMagique** [comment]
Performance may also degrade if you switch data types from row to row in your parameter set, e.g., `float` to `integer`. Any change in parameter types will end the current parameter set.

Best performance is achieved if you do not use "normal" Python objects as parameters, but NumPy arrays or Apache Arrow tables. Check the documentation for `executemanycolumns()`.

**Comment by sevetseh28** [comment]
I rewrote the code to fetch in numpy batches and insert with executemanycolumns. Performance increased a bit but I still get the Exception in another flavor (set_numpy_parameters). Take a look at the new code and the exception:

Code:
```
batches = cursor_azure.fetchnumpybatches()
        unique_words_global = []
        counts_global = []
        insert_cnxn = self.get_azure_conn()
        cursor_insert = insert_cnxn.cursor()
        tweet_ids_global = []
        for batch in batches:
            for idx, tweet_text in enumerate(batch['text']):
                words = text_to_cleaned_words(tweet_text, stopwords_set)
                unique_words, counts = np.unique(words, return_counts=True)

                ids_column = np.full((1, len(unique_words)), batch['id'][idx])[0]

                unique_words_global += (list(unique_words))
                counts_global += (list(counts))
                tweet_ids_global += (list(ids_column))

                if len(tweet_ids_global) > 20000:
                    np_unique_words_global = np.array(unique_words_global, dtype=np.object)
                    np_counts_global = np.array(counts_global, dtype=np.int64)
                    np_tweet_ids_global = np.array(tweet_ids_global, dtype=np.int64)
                    start_time = time.time()
                    cursor_insert.executemanycolumns('INSERT INTO [staging].[tweets_wordcloud] (tweet_id, word, word_count) VALUES (?,?,?)',
                                                    [np_tweet_ids_global, np_unique_words_global, np_counts_global])
                    insert_cnxn.commit()
                    self.logger.info(f'{time.time() - start_time}  secs inserting batch of {len(tweet_ids_global)}')
                    
                    unique_words_global = []
                    counts_global = []
                    tweet_ids_global = []
```

Exception:
```
Traceback (most recent call last):
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/exceptions.py", line 50, in wrapper
    return f(*args, **kwds)
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/cursor.py", line 219, in executemanycolumns
    set_numpy_parameters(self.impl, split_arrays)
turbodbc_intern.Error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hernan/.pycharm_helpers/pydev/pydevd.py", line 1683, in <module>
    main()
  File "/home/hernan/.pycharm_helpers/pydev/pydevd.py", line 1677, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/home/hernan/.pycharm_helpers/pydev/pydevd.py", line 1087, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/hernan/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/mnt/c/IDATHA/Workspace/idatha_sbi/idatha_sbi_etl.py", line 179, in <module>
    dal.load_wordcloud(source=TWITTER)
  File "/mnt/c/IDATHA/Workspace/idatha_sbi/dal/dal.py", line 310, in load_wordcloud
    [np_tweet_ids_global, np_unique_words_global, np_counts_global])
  File "/home/hernan/.pyenv/versions/3.6.4/envs/idatha_sbi/lib/python3.6/site-packages/turbodbc/exceptions.py", line 52, in wrapper
    raise DatabaseError(str(e))
turbodbc.exceptions.DatabaseError
```

**Comment by MathMagique** [comment]
Thanks for the continued effort you put into this! It is much appreciated. I'll try to find the time to dig in further.

**Comment by sevetseh28** [comment]
Hey! I found the bug. It is not a turbodbc bug but a bug in my text to words cleaning method and target database collation.  Since the exception was generic, to find the bug I had to decompose the batch when the Exception occured and within that batch insert one set of words per tweet. I found out that a specific tweet had some strange unicode bold characters:
```
<class 'list'>: ['ACTIVIDADES', 'ENTERATE', 'OBRAS', 'PROYECTOS', '𝗦I𝗴𝘂𝗲𝗻𝗼𝘀', '𝗲𝗻', '𝗻𝘂𝗲𝘀𝘁𝗿𝗮𝘀', '𝗿𝗲𝗱𝗲𝘀', '𝘀𝗼𝗰𝗶𝗮𝗹𝗲𝘀']
```
That was the list of supposed cleaned words for a Tweet.
To make it worse my Azure SQL Database has a **SQL_Latin1_General_CP1_CI_AS** collation so it probably doesn't support those bold unicode characters.

I managed to solve it by normalizing the tweet text in NFKC form:
```
unicodedata.normalize('NFKC', text)
``` 

Hope this will help anyone with this setup!
@MathMagique  thank you so much for your help! I did not only find the bug but also increased performance thanks to your tips 👍

**Comment by MathMagique** [comment]
I'm glad you found the issue, and even more so that it was not in turbodbc ;-). I still wonder why turbodbc did not yield a better error message, as it should forward any errors it gets from the database.

Please allow me two final questions before I close this issue:
* Do you set the `prefer_unicode` option to `True` when establishing the connection, as suggested in the [turbodbc documentation](http://turbodbc.readthedocs.io/en/latest/pages/databases/mssql.html#recommended-turbodbc-configuration)? If not, this might resolve encoding-related issues.
* Do you have some rough indicator of how performance compares to ceODBC?

**Comment by sevetseh28** [comment]
- I did set prefer_unicode to True the whole time. 
- I do not have a rough indicator but I can tell you that ceODBC runs well and fast only with Python 2.7 and on Windows. Even so turbodbc is still faster than ceODBC and now that I rewrote my code it is still much faster and besides it runs on Linux with Python 3.6.4 and works well with the official Microsoft Driver so point for turbodbc in all aspects!

Thanks!
